text/3668-async-closure.md

```diff
@@ -0,0 +1,541 @@
+
+In order for this to work, the `FnMut` trait would need to be ["lending"](https://smallcultfollowing.com/babysteps/blog/2023/05/09/giving-lending-and-async-closures/#async-closures-are-a-lending-pattern); however, there are [complications](https://hackmd.io/@compiler-errors/async-closures#Lending-closures-are-not-typically-FnOnce) with implementing general lending closures.
+
+# Guide Level Explanation
+
+Just as you can write functions which accept closures, you can write functions which accept async closures:
+
+```rust
+async fn takes_async_closure(f: impl async Fn(u64)) {
+    f(0).await;
+    f(1).await;
+}
+
+takes_async_closure(async |i| {
+    core::future::ready(i).await;
+    println!("done with {i}.");
+});
+```
+
+We recommend using `async Fn` and `async ||` for async closures. This is more flexible than a closure returning a future for the reasons described elsewhere in this RFC.
+
+Async closures act similarly to closures, and can have parts of their their signatures specified:
+
+```rust
+// They can have arguments annotated with types:
+let arg = async |x: i32| { async_add(x, 1).await };
+
+// They can have their return types annotated:
+let ret = async || -> Vec<Id> { async_iterator.collect().await };
+
+// They can be higher-ranked:
+let hr = async |x: &str| { do_something(x).await };
+```
+
+When called, they return an anonymous future type corresponding to the (not-yet-executed) body of the closure. These can be awaited like any other future.
+
+# Detailed Explanation
+
+### `AsyncFn*`
+
+This RFC introduces a family of `AsyncFn` traits. These traits are intended to remain perma-unstable to name or implement, just like the `Fn` traits. Nonetheless, we'll describe the details of these traits so as to explain the user-facing features enabled by them.
+
+The definition of the traits is (modulo `rustc_` attributes, and the `"rust-call"` ABI):
+
+
+> [!NOTE]
+> We omit some details about the `"rust-call"` calling convention and the fact that the `Args` parameter is enforced to be a tuple.
+
+```rust
+/// An async-aware version of the [`FnOnce`](crate::ops::FnOnce) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnOnce::async_call_once`].
+    type CallOnceFuture: Future<Output = Self::Output>;
+
+    /// Output type of the called closure's future.
+    type Output;
+
+    /// Call the [`AsyncFnOnce`], returning a future which may move out of the called closure.
+    fn async_call_once(self, args: Args) -> Self::CallOnceFuture;
+}
+
+/// An async-aware version of the [`FnMut`](crate::ops::FnMut) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnMut<Args>: AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnMut::async_call_mut`] and [`AsyncFn::async_call`].
+    type CallRefFuture<'a>: Future<Output = Self::Output>
+    where
+        Self: 'a;
```

Comment:
This is the most surprising part of the RFC to me, and it is not discussed: why do the Fn/FnMut variants return a different type than the FnOnce variant of the trait? That's very inconsistent with the existing `Fn*` family of traits. Won't that cause issues?

One could imagine something like AyncFnOnce returning `CallFuture<'static>`. I don't expect that to work, but the RFC should spend some time discussing this surprising design choice.

If they need different return types, why do AsyncFn and AsyncFnMut still share the same type? Maybe they should be different, too?

Reply:
+1. 

The `Fn : FnMut : FnOnce` hierarchy make sense because one can implement forward the parent `call` method to the child:

```rust
fn call_once(mut self, args: Args) -> Self::Output {
    self.call_mut(args)
}
fn call_mut(&mut self, args: Args) -> Self::Output {
    self.call(args)
}
```

I can't make the same for `AsyncFnMut : AsyncFnOnce` because of that lifetime. 

Following the graph in *Why do we need a new set of `AsyncFn*` traits?*, `AsyncFnMut` is actually a `LendingFnMut<Output: Future>` which should have no isa relationship with `AsyncFnOnce` (= `FnOnce<Output: Future>`).

Reply:
I detail some of this relationship on a "blog" "post" I wrote about async closures: https://hackmd.io/@compiler-errors/async-closures

I'll try to paint it in broad strokes in the RFC but I can answer a few questions here.

> One could imagine something like AyncFnOnce returning CallFuture<'static>. I don't expect that to work, but the RFC should spend some time discussing this surprising design choice.

Yeah that very much doesn't work haha. I kinda try to demonstrate that in the blog post but I can try to boil it down to something that isn't just "the lifetimes don't work out that way lol". I'm mostly just worried that if I go into extreme technical detail I risk confusing people further, especially because the user-facing consequences are not incredibly appreciable.

> If they need different return types, why do AsyncFn and AsyncFnMut still share the same type? Maybe they should be different, too?

Nope -- the salient distinction is whether the closure was called by-ref or by-move since that's what has consequences on the upvars of the returned future. There's never a case where an async closure needs to distinguish its by-mut-ref and by-ref futures, since if an async closure *has* an AsyncFn implementation it mustn't actually take a mutable ref any of its captures.

> AsyncFnMut is actually a LendingFnMut<Output: Future> which does not show any isa relationship with AsyncFnOnce (= FnOnce<Output: Future>).

It's more proper to say that an `AsyncFnMut<Output = T> = for<'s> LendingFnMut<Output<'s>: Future<Output = T>> + FnOnce<Output: Future<Output = T>>`.

So in general, lending closures cannot be called once. This is a property that is recovered by async closures and is re-imposed by the the set of async traits having the shape it has today, but can also just be imposed by expanding `async FnMut() -> T` into both a lending and fnonce bound as I did just above.

It's IMO a desirable thing to be able to call async closures once and be able pass them to (async) FnOnce callsites even if their bodies capture references to their state, so that's why this hierarchy has an additional edge connecting LendingFnMut back to FnOnce even if that's not a universal truth of the traits in general.

> That's very inconsistent with the existing Fn* family of traits. Won't that cause issues?

I'm not totally certain why you perceive this to be *very* inconsistent, or why you perceive this would cause issues -- certain implementations can choose to return the same future for CallRefFuture/CallMutFuture, but in general unless the compiler has hard-coded an assumption that the return type of AsyncFnOnce::call_once and AsyncFnMut::call_mut are the same, I can't really see a problem would arise.

The compiler synthesizes an AsyncFnOnce implementation for the closure just like it does for FnOnce for call-by-{mut-,}-ref regular closures. While this implementation is a bit more involved (we need to generate a new future that takes ownership of all the captures that would be dropping at the end of the AsyncFnOnce) this is just another compiler detail--it's probably a useful detail to mention, since it has (somewhat niche in practice, but still) user-facing semantics about drop ordering and outlives and stuff, but after having written this closure implementation for the last 6 months, I don't really know what people care to know more about vs don't care about it.

Perhaps I should just permalink to (current) internal documentation about how this stuff works rather than repeat myself in the RFC.

The most important part is that these futures are still enforced to have the same output type after being awaited -- if that weren't enforced then I would say these traits are inconsistent, but they'd also be unusable.

Reply:
So compiling an async closure actually generates *two* functions, and the body of the closure gets compiled twice -- once to implement AyncFnOnce, and once for AsyncFn? (And then AsyncFnMut can call the AsyncFn implementation.) Yes that should definitely be mentioned, it makes for a very different mental model than what I had coming in.

Are there situations where one of these would successfully compile but the other does not?

Reply:
> So compiling an async closure actually generates two functions, and the body of the closure gets compiled twice -- once to implement AyncFnOnce, and once for AsyncFn?

We do end up with two MIR bodies that may be monomorphized depending on if you call `AsyncFn`/`AsyncFnMut` or `AsyncFnOnce`. The "CallOnceFuture" body is derived using the [`ByMoveBody`](https://github.com/rust-lang/rust/blob/7d97c59438e933e86f557ed999da3b8dfc6855a7/compiler/rustc_mir_transform/src/coroutine/by_move_body.rs#L4) transform. It performs the proper adjustments to the body to take the upvars by move.

> Are there situations where one of these would successfully compile but the other does not?

Not that I am aware of.

Reply:
> https://hackmd.io/@compiler-errors/async-closures

It'll be good if you can put this in the description for reference :)

Reply:
Okay so for general closures I would question whether `trait LendingFnMut<A>: FnOnce<A>` makes sense -- those seem like basically unrelated traits at that point.

For async closures though they end up being tied together by the output type. And since one usually doesn't care much for the actual concrete type of the future, only the output type, the fact that `AsyncFnOnce` and `AsyncFnMut` get to that output type via completely unrelated futures... basically doesn't matter.

Reply:
I've tried to detail this, but let me know if you want anything in particular spelled out. The problem is that motivating the *why* is incredibly annoying without going through all that could go bad, and I feel like it's not the best place to put this in an RFC. Knowing specifics about what you want elaborated may help, if you're still unsatisfied with the justification here :)

Reply:
That helps, thanks. :)